R/calculate_tdm.R

```diff
@@ -0,0 +1,305 @@
+#' Calculate and add the transition disruption metric to a dataset
+#'
+#' This function takes a typical PACTA output, and uses it to calculate the
+#' transition disruption metric.
+#'
+#' @param data A dataset, like the `*_results_portfolio.rda` outputs of PACTA.
+#' @param t0 The start year, from which the TDM value should be calculated.
+#' @param t1 The number of years into the future for which there is production
+#'   data. Default is 5 years.
+#' @param t2 The number of years into the future against which you want to
+#'   calculate disruption. Default is 10 years.
+#' @param additional_groups Character vector. The names of columns to group by,
+#'   in addition to these ones (which are always used):
+#'
+#'   `r toString(crucial_tdm_groups())`
+#'
+#' @return A tibble with the columns specified in the `...` input, as well as:
+#'   `tdm_tech`, the technology level transition disruption metric and
+#'   `tdm_sec`, the sector level transition disruption metric.
+#' @export
+#'
+#' @examples
+#' library(tibble)
+#'
+#' pacta_results <- tibble(
+#'   investor_name = rep("some_investor", 3),
+#'   portfolio_name = rep("some_portfolio", 3),
+#'   allocation = rep("portfolio_weight", 3),
+#'   equity_market = rep("GlobalMarket", 3),
+#'   scenario_geography = rep("Global", 3),
+#'   ald_sector = rep("Power", 3),
+#'   technology = rep("RenewablesCap", 3),
+#'   year = c(2020, 2025, 2030),
+#'   plan_alloc_wt_tech_prod = rep(1, 3),
+#'   scen_alloc_wt_tech_prod = 1:3,
+#'   plan_carsten = rep(0.5, 3)
+#' )
+#' pacta_results
+#'
+#' calculate_tdm(pacta_results, t0 = 2020)
+calculate_tdm <- function(data, t0, t1 = 5, t2 = 10, additional_groups = NULL) {
+  stopifnot(
+    is.data.frame(data),
+    is.numeric(t0),
+    is.numeric(t1),
+    is.numeric(t2)
+  )
```

Comment:
I'm becoming fond of vctrs because it is rigurous and has nice error messages

``` r
# OK
t0 <- 2020
vctrs::vec_assert(t0, numeric(), size = 1L)

# Nice error message
t0 <- "2020"
vctrs::vec_assert(t0, numeric(), size = 1L)
#> Error: `t0` must be a vector with type <double>.
#> Instead, it has type <character>.

# Nice error message
t0 <- c(2020, 2021)
vctrs::vec_assert(t0, numeric(), size = 1L)
#> Error: `t0` must have size 1, not size 2.

# Is numeric the right prototpye?
t0 <- 2020L
vctrs::vec_assert(t0, numeric(), size = 1L)
#> Error: `t0` must be a vector with type <double>.
#> Instead, it has type <integer>.
```

<sup>Created on 2021-11-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Reply:
until [this issue](https://github.com/2DegreesInvesting/create_interactive_report/issues/422) is resolved, we won't be able to use the `vctrs` package or any package that is not already in the Docker image

Reply:
and in general, I like not adding new dependencies to production code unless there is a _really_ big benefit to them.

Reply:
to be clear, it's not really the create_interactive_report issue, but that issue makes updating the base Docker image we use problematic, which makes adding a new package dependency to it problematic... not impossible, but dicey

Reply:
this seems interesting, but for the purpose of this PR I'm not going to go into the pros/ cons of `vctrs`. Lets save it for another discussion :-)